ssl/statem/statem_lib.c

```diff
@@ -2403,3 +2403,149 @@ int tls13_restore_handshake_digest_for_pha(SSL *s)
     }
     return 1;
 }
+
+#if !defined(OPENSSL_NO_COMP)
+int tls13_construct_compressed_certificate(SSL *s, WPACKET *pkt, CERT_PKEY *cpk,
+                                           uint8_t *context, size_t context_len)
+{
+    WPACKET tmppkt;
+    BUF_MEM *buf = NULL;
+    int ret = 0;
+    size_t length;
+    size_t max_length;
+    COMP_METHOD *method;
+    COMP_CTX* comp = NULL;
+    int comp_len;
+
+    if (s->ext.compress_certificate_rx == TLSEXT_comp_cert_none
+            || (buf = BUF_MEM_new()) == NULL
+            || !WPACKET_init(&tmppkt, buf)) {
+        goto err;
+    }
+
+    /* Use the |tmppkt| for the to-be-compressed data */
+    if (context == NULL) {
+        /* no context available, add 0-length context */
+        if (!WPACKET_put_bytes_u8(&tmppkt, 0))
+            goto err;
+    } else if (!WPACKET_sub_memcpy_u8(&tmppkt, context, context_len))
+        goto err;
+
+    if (!ssl3_output_cert_chain(s, &tmppkt, cpk)) {
+        /* SSLfatal() already called */
+        goto out;
+    }
+
+    /* continue with the real |pkt| */
+    if (!WPACKET_put_bytes_u16(pkt, s->ext.compress_certificate_rx)
+            || !WPACKET_get_total_written(&tmppkt, &length)
+            || !WPACKET_put_bytes_u24(pkt, length))
+        goto err;
+
+    /*
+     * Uncompressibility expansion:
+     * Brotli: per RFC7932: N + 5 + 3 * (N >> 16)
+     * ZLIB: N + 11 + 5 * (N >> 14)
+     * ZSTD: N + 4 + 14  \+ 3 * (N >> 17) + 4
+     */
+    switch (s->ext.compress_certificate_rx) {
+# ifdef ZLIB
+    case TLSEXT_comp_cert_zlib:
+        max_length = length + 11 + 5 * (length >> 14);
```

Comment:
A valid point. But approximate size is "not good enough". The COMP API does not provide a mechanism to calculate the approximate size and not all algorithms support that.

That being said, the handshake is already at a spot where we've determined that the certificate is to be compressed, there's no good way to back out (the state machine really doesn't permit changing the message type while constructing a message).

CPU time would be wasted in determining the size of the compression first, then doing the actual compression. In effect, you would be doubling the CPU time by calculating the size first.

Reply:
Maybe send an Alert message and drop connection in case of compressed size bigger than uncompressed. It's strange behaviour when you expecting a decreasing size of the message but got a message twice bigger than uncompressed. It will be really difficult to catch this problem without alerting from the server-side.

Reply:
I consider that bad behavior. If it can work, it should work. Dropping the connection is a poor substitute for properly handling compression expansion. I would prefer some other diagnostic that indicates that compression expanded.

Reply:
Maybe store somewhere information about the size of compressed, uncompressed and compression algorithm id? Somewhere in SSL structure. With this functionality, it could be possible to track size and add alerts on the server-side. Furthermore, it will be possible to graph the distribution of users with and without compression algorithms.

Reply:
That seems a bit out of scope for the OpenSSL library, beyond simple per-connection stats (but, OpenSSL is not known for it's plethora of connection stats).

Reply:
Unfortunately without any information about compressed certificate size, there is no way to detect the problem. BoringSSL uses callbacks for compression algorithms. This could be used as external cache storage and statistics about clients.

Reply:
A callback API was added